### ISSUE TITLE
Slight visual shift of some UI components

### ISSUE THREAD

**Issue by stemilymill** [phetsims/ph-scale#226]
**Test device**
Dell laptop, Chromebook

**Operating System**
Win10, Chrome OS

**Browser**
Chrome 91.0.4472.124

**Problem description**
For https://github.com/phetsims/qa/issues/669.
In the Micro screen, changing the volume of liquid in the beaker causes some of the numbers in the graph to move slightly up or down, and in the My Solution screen, when the graph is displaying Quantity, setting pH to minimum or maximum value causes the graph, beaker, and beaker control panel to move up or down.

**Steps to reproduce**
Micro screen: 
Alternate adding and removing water from the beaker. 

My Solution screen:
Toggle graph to show Quantity.
Drag the blue OH- or red H3O+ indicator to the top or bottom of the graph.

**Visuals**

Micro screen
![graph and beaker jump micro screen](https://user-images.githubusercontent.com/85511101/126214666-2104685a-3705-4df4-91fc-a8ac667a34cc.gif)
https://drive.google.com/file/d/10dzmjeuit3-fbPRRqMOQ85Sr18qy_sJ4/view?usp=sharing

My Solution screen
![graph and beaker jump my solution screen](https://user-images.githubusercontent.com/85511101/126214680-29528249-4b49-47c1-ae0b-7ab61a3e9b96.gif)

Troubleshooting information:
!!!!! DO NOT EDIT !!!!!
Name: ‪pH Scale‬
URL: https://phet-dev.colorado.edu/html/ph-scale/1.5.0-rc.1/phet/ph-scale_all_phet.html
Version: 1.5.0-rc.1 2021-07-13 18:33:46 UTC
Features missing: applicationcache, applicationcache, touch
Flags: pixelRatioScaling
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36
Language: en-US
Window: 1418x656
Pixel Ratio: 1.3541666269302368/1
WebGL: WebGL 1.0 (OpenGL ES 2.0 Chromium)
GLSL: WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor: WebKit (WebKit WebGL)
Vertex: attribs: 16 varying: 30 uniform: 4096
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 32)
Max viewport: 32767x32767
OES_texture_float: true
Dependencies JSON: {}

**Comment by pixelzoom** [comment]
All of these elements have static positions.  So this is probably a browser problem that I have no control over.   Assigning to @jonathanolson to comment with high priority.

**Comment by pixelzoom** [comment]
@stemilymill Your test device might be behind a few versions. The latest version of Chrome is Chrome 91.0.4472.124.  Can you please update and re-test, to verify that this is not a bug that has been fixed in the browser?

**Comment by pixelzoom** [comment]
@stemilymill Your test device is behind a few versions. The latest version of Chrome is 91.0.4472.164. Can you please update and re-test, to verify that this is not a bug that has been fixed in the browser?

**Comment by stemilymill** [comment]
@pixelzoom My device is now up-to-date and this all behaves exactly the same.

**Comment by pixelzoom** [comment]
@stemilymill Got it, thanks for checking.

**Comment by jonathanolson** [comment]
So far I'm unable to reproduce (Win10, Chrome 91.0.4472.164). It looks potentially size-specific, so I matched the browser size with the reported resolution (1418x656) and still didn't see it.

I did see an interesting pixel ratio in the debug info (`Pixel Ratio: 1.3541666269302368/1`), there definitely could be something going on specific to that setup on the browser-side.

@pixelzoom could we try a test version where we have a `layerSplit: true` separating the graph background (the shifting text instances, labels, etc.), so that the foreground controls/readouts are in a separate layer? @stemilymill do you know if this is broken on master (e.g. on phettest)?

**Comment by pixelzoom** [comment]
> @pixelzoom could we try a test version where we have a `layerSplit: true` separating the graph background (the shifting text instances, labels, etc.), so that the foreground controls/readouts are in a separate layer?

Please clarify.  Set `layerSplit: true`  for the graph background, or for the elements that are shifting around?

Since the entire beaker and panel below the beaker are also shifting around, would it also be sufficient to set `layerSplit: true` for those entire elements?  That might be easier than trying to separate out parts of the graph.

**Comment by jonathanolson** [comment]
I'd like the layerSplit to be put on something that:

1. Contains all of the text that is shifting, but should not be shifting. Can contain other things
2. Does NOT include things that are actually moving, like the indicators or things in the play area.

I'm curious if that will prevent the bug from happening. Ideally do this with one layerSplit... one per tick would probably be bad for performance.

**Comment by pixelzoom** [comment]
Discussed with @jonathanolson on Zoom. The plan is:

* Publish a dev version from master, see if @stemilymill can reproduce it there. If this is reproducible in master, that makes our lives quite a bit easier.
* If reproducible in master, add `layerSplit: true` to the 2 draggable indicators, publish another dev version, and ask @stemilymill to compare behavior.
* Based on the visuals, elements appear to shift when the indicator is dragged to it max value. Verify whether that's actually the case, or whether there is shifting as the indicator is dragged within a smaller range.  If it's specific to max, the fact that the displayed value changes from scientific notation to  non-scientific notation may involve a resize that is significant.

**Comment by pixelzoom** [comment]
@emily-phet two requests:

(1) Please test [1.6.0-dev.2](https://phet-dev.colorado.edu/html/ph-scale/1.6.0-dev.2/phet/ph-scale_en_phet.html) to see if the problem occurs in that version.  (This is a dev version built from master, with no changes/fixes.  If it's reproducible there, then we can troubleshoot in master, rather than in the 1.5 branch.)

(2) Based on the visuals that you provided, UI elements appear to shift when the indicator is dragged to its max value (e.g. OH- = 5.0).  Is that a requirement to make this happen, or does the shifting occur as you drag the indicator through a smaller range? 

Thanks!

**Comment by emily-phet** [comment]
I think the comment above was intended for @stemilymill.

**Comment by pixelzoom** [comment]
Good news -- I can reproduce this in master on macOS 11.3.1 + Chrome 91.0.4472.164.  The key is indeed to move drag the indicator to its maximum. 

Steps to reproduce:
1. Run the sim from master
2. Go to the My Solution screen
3. Set the AB switch at the top of the graph to "Quantity".
4. Drag the OH- indicator to its max value (5.00).  Then alternate between 5.00 and just below 5.00.  You'll see several UI elements shift up and down: the beaker, the panel below the beaker, and the entire body of the graph.

**Comment by pixelzoom** [comment]
@stemilymill you can ignore the requests in https://github.com/phetsims/ph-scale/issues/226#issuecomment-883570502, since I've reproduced this on macOS in master.

**Comment by pixelzoom** [comment]
I tried adding `layerSplit: true` to the default options for `GraphIndicatorNode`.  Running unbuilt from master, this resulted in no change in behavior. Static UI elements continue to shift up and down when following the steps in https://github.com/phetsims/ph-scale/issues/226#issuecomment-883576284.

@jonathanolson thoughts on how to proceed?  And can you also reproduce this now, following the steps in https://github.com/phetsims/ph-scale/issues/226#issuecomment-883576284?

**Comment by pixelzoom** [comment]
> If it's specific to max, the fact that the displayed value changes from scientific notation to non-scientific notation may involve a resize that is significant.

This doesn't appear to be related to resizing the value display. I added this to GraphIndicatorNode, to constrain the size:

```js
    valueNode.maxWidth = 100;
    valueNode.maxHeight = valueNode.height;
```

... and there was no change in behavior.

**Comment by arouinfar** [comment]
This reminds me of https://github.com/phetsims/gravity-force-lab/issues/101, which is currently in the published version of Gravity Force Lab. The underlying cause may be different, but there's certainly precedent for publishing a sim with sort of visual wiggle.

This issue occurs only in a very specific situation and doesn't affect usability. Given that there isn't an obvious solution, I'm fine with deferring or closing as wont-fix. Your call @pixelzoom.

**Comment by jonathanolson** [comment]
Unable to reproduce, moving the OH indicator here:

![image](https://user-images.githubusercontent.com/480878/126674389-249ac958-3a09-4014-8e0f-6e8fe2552b53.png)

@pixelzoom what if you put the flag `preventFit: true` on the parts that are shifting (with no layerSplit)?

**Comment by pixelzoom** [comment]
> Unable to reproduce, moving the OH indicator here:

You haven't moved it to its maximum, 5.0. Your screenshot shows 4.8. 

> @pixelzoom what if you put the flag preventFit: true on the parts that are shifting (with no layerSplit)?

I'll give it a try.

**Comment by jonathanolson** [comment]
Understood, I don’t have a good setup for taking videos, so I wanted to make it clear that I was moving it “from” the 5.0 position. I moved it from the top to where the screenshot was repeatedly, and tried other combinations.

**Comment by pixelzoom** [comment]
@jonathanolson are you on macOS + Chrome Chrome 91.0.4472.164?  Odd that I can reproduce it 100% of the time and you can't.

**Comment by jonathanolson** [comment]
![image](https://user-images.githubusercontent.com/480878/126708845-3ed5fd55-9c86-4c1c-bb40-1813f4633abb.png)
![image](https://user-images.githubusercontent.com/480878/126708861-2fde933d-5da9-4c55-b2fd-7661380ee936.png)

No shift noticed, screenshots in both conditions. macOS 11.4, Chrome 91.0.4472.114, running from master, steps taken from https://github.com/phetsims/ph-scale/issues/226#issuecomment-883576284.

**Comment by jonathanolson** [comment]
Updated to Chrome 92.0.4515.107 (macOS), still no reproduction of the bug, either in the full-screen or with devtools open.

**Comment by pixelzoom** [comment]
Removing high-priority label, since @arouinfar says that this is non-blocking, and could even be closed as "won't fix".

**Comment by pixelzoom** [comment]
I upgraded Chrome to 92.0.4515.107 (lastest). With macOS 11.3.1, there was no change.

Then I upgraded macOS to 11.5 (latest). With Chrome to 92.0.4515.107, there was again no change.

So I'll move on to trying `preventFit: true`.

**Comment by pixelzoom** [comment]
In the My Solution screen, there's a lot more shifting around than I originally notices. Here's the (complete?) list of things that shift up when the steps in https://github.com/phetsims/ph-scale/issues/226#issuecomment-883576284 are followed:

These parts of ph-scale/MySolutionScreenView:
* The entire control panel below the beaker, `beakerControlPanel`
* The entire beaker, `beakerNode`
* The solution in the beaker, `solutionNode`
* The molecule counts, `moleculeCountsNode`
* The Reset All button, `resetAllButton`

These parts of ph-scale/GraphNode:
* The complete graph body, incuding tick marks and labels
* The value in the H2O indicator on the graph, but not the rest of the H2O indicator

There parts of ph-scale/PHMeterNode:
* The "pH" label in the pH meter, which is the AccordionBox `titleNode` option
* The pH probe, `probeNode`.

If I set ANY NODE ON THIS SCREEN to `preventFit: true`, then this problem goes away.  It doesn't have to be one of the Nodes listed about that is shifting around.  For example, I add `preventFit: true` to `resetAllButton` in MySolutionScreenView.js, one of the Nodes that is shifting around.  Or I can add `preventFit: true` to something like `pHMeterNode` in PHMeterNode.js, one of the Nodes that is NOT shifting around.  All it takes is for one Node on the screen to have `preventFit: true`, and the problem disappears.

@jonathanolson does that tell you anything about the nature of the problem?  And do you think it would be OK to just add `preventFit: true` to `resetAllButton` on both screens as a workaround?

**Comment by pixelzoom** [comment]
I verified that adding `preventFit: true` to `resetAllButton` in MicroScreenView.js also fixes the problem on that screen.

**Comment by pixelzoom** [comment]
I'm not familiar with `preventFit`, so here's the doc from Node.js:

```js
      // {boolean} - If true, Scenery will not fit any blocks that contain drawables attached to Nodes underneath this
      //             Node's subtree. This will typically prevent Scenery from triggering bounds computation for this
      //             sub-tree, and movement of this Node or its descendants will never trigger the refitting of a block.
      preventFit: DEFAULT_OPTIONS.preventFit
```

So this sounds like a safe thing to set on a screen's ResetAllButton.  Is that correct?

**Comment by jonathanolson** [comment]
> So this sounds like a safe thing to set on a screen's ResetAllButton, since it never moves. Is that correct?

It's a safe thing in general, mainly could only reduce/increase performance.

> @jonathanolson does that tell you anything about the nature of the problem? And do you think it would be OK to add preventFit: true to resetAllButton on both screens as a workaround?

On the ResetAllButton sounds good, but placing preventFit on the content that moves might be more appropriate, in case layer splits are added in the future. Currently, it seems everything is going into one layer, so the preventFit flag on any of those nodes will cause the improved behavior.

Presumably the shifting of the SVG element (being fitted) was causing a number of problems, so I'm curious to see if other issues are noticed in other sims as well. `preventFit` mainly exists to reduce memory usage and improve performance, so if it starts to wreck sims then we could presumably remove it.

**Comment by pixelzoom** [comment]
Thanks @jonathanolson.

I'm going to go with the least-cost solution here, and patch it into the 1.5 branch.  I'll  put `preventFit: true` on the ResetAllButton instances, along with a comment referencing this issue, and noting that this should be revisited if layer splits are added in the future.

**Comment by pixelzoom** [comment]
Patched in master and 1.5 branches.  While this problem doesn't appear in ph-scale-basics, I did patch it in the 1.5 branch to prevent possible merge conflicts with future changes.

Ready for verification in the next RC.   

Steps to verify for the My Solution screen:

1. Go to the My Solution screen
2. Set the AB switch at the top of the graph to "Quantity". (Does NOT happen with switch set to "Concentration"!)
3. Drag the OH- indicator to its max value (5.00). Then alternate between 5.00 and just below 5.00. You should not see any UI elements shifting up and down.

Steps to verify for the Micro screen:

1. Go to the Micro screen
2. Set the solute combo box to "Battery Acid" 
3. Change the volume of solution in the beaker by alternative between filling the beaker using the water faucet, and draining solution using the drain faucet.   You should not see any UI elements shifting up and down.

**Comment by pixelzoom** [comment]
After a discussion with @jonathanolson and learning more in  phetsims/scenery#1289, I moved the `preventFit: true` workaround to the ScreenView level.  It's now applied to all ScreenViews in this sim via this constant in PHScaleConstants:

```js
  SCREEN_VIEW_OPTIONS: {
    layoutBounds: new Bounds2( 0, 0, 1100, 700 ),

    // Workaround for things shifting around while dragging
    // See https://github.com/phetsims/scenery/issues/1289 and https://github.com/phetsims/ph-scale/issues/226
    preventFit: true
  },
```
